### PR TITLE
added documentation about zope4 ZMI template updating

### PR DESCRIPTION
**Pull request by drfho**
please check comprehensibility, thanks Frank

### CODE REVIEW

File: docs/ZMI.rst [hunk 1]
```diff
@@ -98,3 +98,157 @@ The properties can have one of the following types:
...
+ZPT-Example see: ../Zope/src/Products/PageTemplates/www/ptEdit.zpt
+
+**Hints and Warnings:** 
+Some input field show additional informations; these are added as element 
+``<small>`` directly following the referred inout field. (Both element are nested 
+by the width defining div-container). Possible text colors are declared by 
+typical bootstrap class names like ``text-warning``.
+
+**Icons:** 
+Zope4 object classes which are show in the ZMI have declared a global variable
```
Comment (dwt) [inline docs/ZMI.rst]: My understanding is that these are instance variables? Is Zope Speak here different to standard python?

File: docs/ZMI.rst [hunk 2]
```diff
@@ -98,3 +98,157 @@ The properties can have one of the following types:
...
+**Textarea:** 
+A textarea element for editing template or script code uses the JS library 
+``ace`` for syntax high-lighting and line numbering. Textarea elements which 
+are declared by the CSS class ``zmi-code`` are transformed into an ace-edditor 
+field. Moreover this element has an attribute ``data-contenttype`` which is 
+needed by ace-editor to determine the fitting syntax high-lighting.
+ZPT-Example see: ``../Zope/src/Products/PageTemplates/www/ptEdit.zpt``
+
+**File upload element:** 
+The file upload element has an own form container (classfied as ``zmi-upload``).
```
Comment (dwt) [inline docs/ZMI.rst]: Maybe "**it's** own form container" could be more fluid.

File: docs/ZMI.rst [hunk 3]
```diff
@@ -98,3 +98,157 @@ The properties can have one of the following types:
...
+	</div
+
+The following example code shows a whole restructed DTML template rendering the 
+Zope4 ZMI
+
+**Example: updated DTML template** 
+(from: ``../Zope/src/OFS/dtml/documentEdit.dtml``)
+
+.. raw:: html
+	<dtml-var manage_page_header>
```
Comment (dwt) [inline docs/ZMI.rst]: Inlining the file (and thus keeping it simple, even if the actual template becomes more complex) can be valuable - but always showing the actual used code might also be.

If wanted, can Sphinx accept

```
.. raw:: html
   :file: filename.html
```

To include the actual template used?

Not sure this is a good idea.

File: docs/ZMI.rst [hunk 4]
```diff
@@ -98,3 +98,157 @@ The properties can have one of the following types:
...
+**Icons:** 
+Zope4 object classes which are show in the ZMI have declared a global variable 
+``zmi_icon``; this string corresponds to an appropiate font icon-CSS class 
+supplied by the Fontawsome web font (https://fontawesome.com/icons)
+
+**Tables:** 
+Bootstraps requires an explictit CSS class ``table`` for any table; espcially 
+long item lists should get an additional CSS class ``table-sm`` and maybe another 
+class ``table-striped`` for a better readability. Finally it is recommended 
+to add a specific identifying class name. The table structure is
```
Comment (dwt) [inline docs/ZMI.rst]: This seems to be missing part of the sentence

File: docs/ZMI.rst [hunk 5]
```diff
@@ -98,3 +98,158 @@ The properties can have one of the following types:
...
+	<div class="form-group row">
+		<label for="title" class="form-label col-sm-3 col-md-2">Title</label>
+		<div class="col-sm-9 col-md-10">
+			<input id="title" class="form-control" type="text" name="title" 
+				value="<dtml-if title>&dtml-title;</dtml-if>" />
+		</div>
+	</div>
+
+The following buttons are constructed as div element with the classname 
+``zmi-controls``; the bbuttons use systematically the  bootstrap class pair
```
Comment (sallner) [inline docs/ZMI.rst]: _b_uttons + whitespace later on

File: docs/ZMI.rst [hunk 6]
```diff
@@ -98,3 +98,158 @@ The properties can have one of the following types:
...
+	</div>
+
+The following buttons are constructed as div element with the classname 
+``zmi-controls``; the bbuttons use systematically the  bootstrap class pair 
+``btn btn-primary``.
+
+.. raw:: html
+	<div class="zmi-controls">
+		<input class="btn btn-primary" type="submit" name="submit" value="Save" />
+	</div
```
Comment (sallner) [inline docs/ZMI.rst]: This div should be closed.

File: docs/ZMI.rst [hunk 7]
```diff
@@ -98,3 +98,158 @@ The properties can have one of the following types:
...
+	
+	<dtml-var manage_page_footer>
+
+More details
+------------
+
+**Textarea:** 
+A textarea element for editing template or script code uses the JS library 
+``ace`` for syntax high-lighting and line numbering. Textarea elements which 
+are declared by the CSS class ``zmi-code`` are transformed into an ace-edditor
```
Comment (sallner) [inline docs/ZMI.rst]: ace-e_d_itor

File: docs/ZMI.rst [hunk 8]
```diff
@@ -98,3 +98,158 @@ The properties can have one of the following types:
...
+field. Moreover this element has an attribute ``data-contenttype`` which is 
+needed by ace-editor to determine the fitting syntax high-lighting.
+ZPT-Example see: ``../Zope/src/Products/PageTemplates/www/ptEdit.zpt``
+
+**File upload element:** 
+The file upload element has it's own form container (classfied as ``zmi-upload``). 
+All subseqeunt elements are nested as 'input-group' containing a div classified as 
+``custom-file`` nestingthe actual input element. An inline JS fired on the 
+onchange-event beautifies the file name showed after selecting it.
+ZPT-Example see: ../Zope/src/Products/PageTemplates/www/ptEdit.zpt
```
Comment (sallner) [inline docs/ZMI.rst]: You used ` ``../Zope/src/Products/PageTemplates/www/ptEdit.zpt`` ` before

File: docs/ZMI.rst [hunk 9]
```diff
@@ -98,3 +98,158 @@ The properties can have one of the following types:
...
+``ace`` for syntax high-lighting and line numbering. Textarea elements which 
+are declared by the CSS class ``zmi-code`` are transformed into an ace-edditor 
+field. Moreover this element has an attribute ``data-contenttype`` which is 
+needed by ace-editor to determine the fitting syntax high-lighting.
+ZPT-Example see: ``../Zope/src/Products/PageTemplates/www/ptEdit.zpt``
+
+**File upload element:** 
+The file upload element has it's own form container (classfied as ``zmi-upload``). 
+All subseqeunt elements are nested as 'input-group' containing a div classified as 
+``custom-file`` nestingthe actual input element. An inline JS fired on the
```
Comment (sallner) [inline docs/ZMI.rst]: nesting_ _the

File: docs/ZMI.rst [hunk 10]
```diff
@@ -98,3 +98,158 @@ The properties can have one of the following types:
...
+
+**File upload element:** 
+The file upload element has it's own form container (classfied as ``zmi-upload``). 
+All subseqeunt elements are nested as 'input-group' containing a div classified as 
+``custom-file`` nestingthe actual input element. An inline JS fired on the 
+onchange-event beautifies the file name showed after selecting it.
+ZPT-Example see: ../Zope/src/Products/PageTemplates/www/ptEdit.zpt
+
+**Hints and Warnings:** 
+Some input field show additional informations; these are added as element
```
Comment (sallner) [inline docs/ZMI.rst]: information_ without s

File: docs/ZMI.rst [hunk 11]
```diff
@@ -98,3 +98,158 @@ The properties can have one of the following types:
...
+**File upload element:** 
+The file upload element has it's own form container (classfied as ``zmi-upload``). 
+All subseqeunt elements are nested as 'input-group' containing a div classified as 
+``custom-file`` nestingthe actual input element. An inline JS fired on the 
+onchange-event beautifies the file name showed after selecting it.
+ZPT-Example see: ../Zope/src/Products/PageTemplates/www/ptEdit.zpt
+
+**Hints and Warnings:** 
+Some input field show additional informations; these are added as element 
+``<small>`` directly following the referred inout field. (Both element are nested
```
Comment (sallner) [inline docs/ZMI.rst]: in_p_ut ... Both element_s_

File: docs/ZMI.rst [hunk 12]
```diff
@@ -98,3 +98,158 @@ The properties can have one of the following types:
...
+A textarea element for editing template or script code uses the JS library 
+``ace`` for syntax high-lighting and line numbering. Textarea elements which 
+are declared by the CSS class ``zmi-code`` are transformed into an ace-edditor 
+field. Moreover this element has an attribute ``data-contenttype`` which is 
+needed by ace-editor to determine the fitting syntax high-lighting.
+ZPT-Example see: ``../Zope/src/Products/PageTemplates/www/ptEdit.zpt``
+
+**File upload element:** 
+The file upload element has it's own form container (classfied as ``zmi-upload``). 
+All subseqeunt elements are nested as 'input-group' containing a div classified as
```
Comment (sallner) [inline docs/ZMI.rst]: subseq_ue_nt ... ` ``input-group`` ` as you used this notation for all CSS-classes

File: docs/ZMI.rst [hunk 13]
```diff
@@ -98,3 +98,158 @@ The properties can have one of the following types:
...
+by the width defining div-container). Possible text colors are declared by 
+typical bootstrap class names like ``text-warning``.
+
+**Icons:** 
+Zope4 object classes which are shown in the ZMI have declared a class variable 
+``zmi_icon``; this string corresponds to an appropiate font icon-CSS class 
+supplied by the Fontawsome web font (https://fontawesome.com/icons)
+
+**Tables:** 
+Bootstraps requires an explictit CSS class ``table`` for any table; espcially
```
Comment (sallner) [inline docs/ZMI.rst]: Bootstrap__ ... explic__it ... esp_e_cially

File: docs/ZMI.rst [hunk 14]
```diff
@@ -98,3 +98,158 @@ The properties can have one of the following types:
...
+
+**Tables:** 
+Bootstraps requires an explictit CSS class ``table`` for any table; espcially 
+long item lists should get an additional CSS class ``table-sm`` and maybe another 
+class ``table-striped`` for a better readability. Finally it is recommended 
+to add a specific identifying class name. The table structure is compliant to 
+bootstrap standard table (https://getbootstrap.com/docs/4.1/content/tables/).
+
+**ZMI-classes:** 
+All basic stylings of the zmi-elements  are defined in the CSS file, see:
```
Comment (sallner) [inline docs/ZMI.rst]: one whitespace ... zmi-elements_ _are

File: docs/ZMI.rst [hunk 15]
```diff
@@ -98,3 +98,158 @@ The properties can have one of the following types:
...
+**Icons:** 
+Zope4 object classes which are shown in the ZMI have declared a class variable 
+``zmi_icon``; this string corresponds to an appropiate font icon-CSS class 
+supplied by the Fontawsome web font (https://fontawesome.com/icons)
+
+**Tables:** 
+Bootstraps requires an explictit CSS class ``table`` for any table; espcially 
+long item lists should get an additional CSS class ``table-sm`` and maybe another 
+class ``table-striped`` for a better readability. Finally it is recommended 
+to add a specific identifying class name. The table structure is compliant to
```
Comment (sallner) [inline docs/ZMI.rst]: can you please give an example for such a name?

**Comment by dwt** [comment]
lgtm

**Pull request merged**